### ISSUE TITLE
Components as event sources

### ISSUE THREAD

**Issue by pmros** [joaomilho/act#31]
I just wanna discuss about component communication in Act. I'll use submission example (recently added to this repo).

In submission example, a component (`submission`) send a "message" (`disabled`) to another component (`confirm`, a children or nested component). And `confirm` component send another message (`confirmSubmit`) to `submission` component.

In order to send `disabled` message, the `submission` component modifies `confirm` component model. In order to send `confirmSubmit` message, `submission` captures an action (`confirmSubmit`), sended by `confirm` component.

I think this communication mechanism is not very clear because:
a) You use different mechanism to send/receive messages (model and actions).
b) There is no clear sepparation beetween which part of component is internal/private (the part that is not interesting for other components) and external/public (relative of communication with other components). I guess most people think about components like black boxes with inputs and outputs.

So I think it would be clearer if components could send signals (side effects), acting like event sources, and they could subscribe from antoher components (because they would be event sources). I think this approach is clearer but it has cons: 
a) It's really not needed as you can use the actual mechanism. 
b) It could be (a bit) slower and more redundant (and maybe more lines of code). It's like to send something that is "inside the world" to "outside the world" (through side effects) and then back to "inside the world" (through subscriptions). It's supposed you have to minimize "outside the world" (side effects are "bad").

I'm not talking about rethinking or remaking Act, it's only another feature (easy to implement, I guess) you may use or not (it's up to you).

We would need a `send` (or `emmit`) additional parameter in sideeffects:
```javascript
const sideEffect = (payload, history, send) => {
    // ...
    send(message)
    // ...
}
```

and a `fromComponent` funtion to convert a component in a event source (like `fromEvents` but for compoenents):
```javascript
const subscribtion = {
    messages: fromComponent(component)
}
```

It's something like choo but with signals (like Angular2).

These are only opinions from a (forever) poor newbie, maybe I'm wrong. But I think discussion is good for the community.
Thank you for your attention.

**Comment by joaomilho** [comment]
Ok, I'm back from CurryOn. Yes, I'm not sure copying Elm is the perfect fit for component composition. I'm also not sure messages are the best way, since it is a pattern that can be easily used to shoot your foot. @xaviervia has plenty of xp on this, maybe he can weight in.

In this specific example, I find it a little bit odd that the inner component would emit `confirmSubmit`, since this component needs to know there's a listener, therefore it is not _really_ self contained. I think it's best if we can settle for an 100% contained way.

This is also not that different from React, except React components can only "communicate" with their immediate parent by callbacks. In this way any component can communicate with any component, and this is what I think can lead to making the situation unmanageable.

Although [as I already said] I don't have a clear solution to this, I have had a thorough discussion with @xaviervia about https://github.com/xaviervia/redux-haiku. I think the current code is inevitable weird, but this is Redux's fault. A port of haiku to act would be much simpler, and could become a standard for side-effects. I suspect that along the way it may also hint to a solution to this specific issue.

**Comment by joaomilho** [comment]
And about your other comments, yeah, Act is not 1.0.0 yet, and if we have to rewrite it to get to what we want, so be it :D

**Comment by joaomilho** [comment]
A final note: Although I see the didactic value of the submission example, in practice I wouldn't decouple these two components. An implementation with them coupled would be so much simpler, and there's no reason to separate them, except the fact the exercise says so :D

**Comment by pmros** [comment]
Coupled implementation is easier but decoupled components are reusables. It's assumed that you are going to use confirm component many times.
By the way, I'm glad that you are back :tada:

**Comment by joaomilho** [comment]
Yeah, you're right. The question here is whether components should communicate via callbacks or through http://guide.elm-lang.org/architecture/index.html (that's what we're basically copying now).

Cycle's way is interesting, and I actually gave it a shot when I started Act, but I think keeping events attached to the view definition is more human friendly. WDYT?

**Comment by pmros** [comment]
I don't know... I was playing with CycleJS... What I like about CycleJS is that it defines a component as: sources (input), main function (black box) and sinks (output). But I feel CycleJS is not very "friendly" or "pragmatic" sometimes. But it could be friendlier with appropiate drivers/helpers ([CycleJS driven by state](http://www.christianalfoni.com/articles/2016_04_06_CycleJS-driven-by-state)).
I don't know if communicating components thorough callback or subscripctions is possible but I like this component structure: input, black-box, output.

**Comment by joaomilho** [comment]
Exactly. That's my feeling about cycle as well. It is great and conceptually clean, but I feel a framework needs to appeal also for newcomers.

The challenge of making Act in this architecture is that we can't have events defined in the view.
